### PR TITLE
[Do Not Merge] Onboard DataVersion and DataContainer

### PR DESCRIPTION
**Pull request by kseager**
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of “no breaking changes
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist
- [x] Ensure to check this box if one of the following scenarios meet updates in the PR, so that label “WaitForARMFeedback” will be added automatically to involve ARM API Review. Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs, all “removals” and “adding a new property” no more require ARM API review.
  - Adding new API(s)
  - Adding a new API version
  - Adding a new service

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If there are following updates in the PR, ensure to request an approval from API Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 

- [ ] Removing API(s) in stable version
- [ ] Removing properties in stable version
- [ ] Removing API version(s) in stable version
- [ ] Updating API in stable version with Breaking Change Validation errors
- [ ] Updating API(s) in preview over 1 year

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

**Comment by markcowl** [comment]
@kseager Please follow the directions in 'How to fix' link above: https://aka.ms/ci-fix  before we can review this PR.

**Comment by kseager** [comment]
@markcowl all gates are passing, please review

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataContainer.json
Comment (markcowl) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataContainer.json:236]: You should have a reference to a resource type that implements the resource manager envelopw, like this: https://github.com/Azure/azure-rest-api-specs/blob/master/specification/compute/resource-manager/Microsoft.Compute/stable/2020-06-30/disk.json#L1414

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataContainer.json
Comment (markcowl) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataContainer.json:263]: this would be part of the resource envelope definition

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataVersion.json
Comment (markcowl) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataVersion.json:181]: Is this (or any of these operations) an LRO / Async Operation?

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataVersion.json
Comment (markcowl) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataVersion.json:262]: See comments on adding a type for the resource envelope as above.  It would be best to have a single resource you can reference from all of these

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataVersion.json
Comment (markcowl) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataVersion.json:354]: You are encouraged to use x-ms-enum with ModelAsString: true, this will prevent new values from causing breaking changes

**Review by markcowl — CHANGES_REQUESTED** [review]
It looks like this is only a partial copy of this api-version.  You will need to have a complete representation of this api in the initial check-in, as additions in thsi same api-version will not be allowed (by rule) after this.  If several people need to work on this api, you might consider asking for a feature branch to do that work in before the api is moved to master

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataVersion.json
Comment (kseager) [reply specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataVersion.json:181]: yes they are async

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataContainer.json
Comment (kseager) [reply specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataContainer.json:236]: the hesitation on that is that the object is empty otherwise

**Comment by kseager** [comment]
@markcowl we are onboarding a bunch of proxy resources and have already synced with ARM folks about this scenario thus the do not merge flag. At this time we just need the ARM time to review our changes, but thank you for the review.

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataContainer.json
Comment (chiragg4u) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataContainer.json]: >data [](start = 145, length = 4)

All the types should be using plural strings, please consider datacontainers.

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/examples/getDataContainer.json
Comment (chiragg4u) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/examples/getDataContainer.json:15]: >"properties": {}, [](start = 8, length = 17)

Please provide example to understand how this will look like

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataContainer.json
Comment (chiragg4u) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataContainer.json]: >DataContainer [](start = 35, length = 13)

Where is this defined?

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataVersion.json
Comment (chiragg4u) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataVersion.json]: >data [](start = 145, length = 4)

Rename the data type to plural string and name more appropriate string

Comment (chiragg4u) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataVersion.json]: >/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.MachineLearningServices/workspaces/{workspaceName}/data [](start = 7, length = 142)

How is this API different from dataContainer API?

Comment (kseager) [reply specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataVersion.json]: DataContainer is a parent class that will end up doing more versioning management

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataContainer.json
Comment (kseager) [reply specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataContainer.json]: this has been updated to data

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataVersion.json
Comment (markcowl) [reply specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataVersion.json:181]: than make sure you mark any such operations with the x-ms-long-running-operation extension: https://github.com/Azure/autorest/blob/master/docs/extensions/readme.md#x-ms-long-running-operation

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataVersion.json
Comment (kseager) [reply specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataVersion.json]: we are using the term data as plural.

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataContainer.json
Comment (kseager) [reply specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataContainer.json]: please take a look at the updated swagger; this is now AssetContainer

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataVersion.json
Comment (kseager) [reply specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataVersion.json]: data is the plural type we are using. Not sure if I'm following the rest of your comment, let me know if you need anything else.

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/examples/getDataContainer.json
Comment (kseager) [reply specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/examples/getDataContainer.json:15]: this is the example for getDataContainer, it returns a named container. It currently has very minimal meta-data at this time.

**Comment by markcowl** [comment]
@kseager You will need to fix (or get signoff from the ARM team and suppress) the linter errors, see: http://aka.ms/ci-fix

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataVersion.json
Comment (ravbhatnagar) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataVersion.json:216]: same comment on parity between request body and response body.

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataVersion.json
Comment (ravbhatnagar) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataVersion.json:310]: DataVersion model and Data model look exactly the same. Can you model these in a way where the common properties between data and dataversion are exposed under the data resource type. and the delta is exposed to dataVersion type. 
Other wise, i dont see value in this versions resource type. Why cant you have just one type? isnt it versioning same as creating a new instance with a different name?

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataVersion.json
Comment (ravbhatnagar) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataVersion.json:181]: Is there any nuance around re-PUT on an existing version?

**Review by ravbhatnagar — COMMENTED** [review]
@kseager - few comments

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataVersion.json
Comment (kseager) [reply specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataVersion.json:181]: if re-PUT the existing version it will do an update on that name+version

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataContainer.json
Comment (kseager) [reply specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataContainer.json:293]: We've already had a discussions with some other ARM folks about the tags property and came to the conclusion of keeping that in our definition

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataVersion.json
Comment (markcowl) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataVersion.json:1]: Just curious why this isn't included in the previous spec - seems like data versions are a direct child of data containers

Comment (kseager) [reply specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataVersion.json:1]: Please add a little more detail, not following the question

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataVersion.json
Comment (pilor) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataVersion.json:185]: If the PUT is long running the resource needs to have a provisioningState property.  See https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/Addendum.md#creatingupdating-using-put

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataVersion.json
Comment (pilor) [inline specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataVersion.json:304]: Is this a resource ID of some other resource?  If so, what permissions does someone need on it to use it in a dataVersion resource and how do you validate that?

**Review by pilor — CHANGES_REQUESTED** [review]

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataVersion.json
Comment (kseager) [reply specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataVersion.json:304]: This is a resourceID of a datastore resource. Permissions are handled by Datastore and should be resolved by the time this ID reaches DataVersion. We validate it by using the error handling built-into the ResourceArmScope parsing.

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataVersion.json
Comment (kseager) [reply specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataVersion.json]: (removed property)

**Review by pilor — APPROVED** [review]
Consider using better examples.  The examples provided in recent machinelearningservices PRs are not useful to customers

File: specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataVersion.json
Comment (markcowl) [reply specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/preview/2020-09-01-preview/dataVersion.json:1]: Just wondering why these are separate files - generally the file split would be at the tracked (top level)  resource level

**Review by markcowl — APPROVED** [review]

**Comment by markcowl** [comment]
@kseager  review is complete - since this is marked 'do not merge' waiting to see what you want to do with this.

**Comment by markcowl** [comment]
@kseager let me know if there is any further action on this.  I will close if by the end of next week if there is no response.

**Comment by markcowl** [comment]
Closing for lack of response

**Pull request closed without merging**
